pkg/agent/multicast/mcast_controller.go

```diff
@@ -0,0 +1,249 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package multicast
+
+import (
+	"context"
+	"net"
+	"sync"
+	"time"
+
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog/v2"
+
+	"antrea.io/antrea/pkg/agent/config"
+	"antrea.io/antrea/pkg/agent/interfacestore"
+	"antrea.io/antrea/pkg/agent/openflow"
+)
+
+type eventType uint8
+
+const (
+	groupJoin eventType = iota
+	groupLeave
+
+	// Use IGMPv3 for IGMP query message because it is the main version for IGMP.
+	// TODO: support sending IGMP query with both v2 and v3 messages
+	queryVersion = 3
+
+	workerCount = 2
+
+	podInterfaceIndex = "podInterface"
+)
+
+type mcastGroupEvent struct {
+	group net.IP
+	eType eventType
+	time  time.Time
+	iface *interfacestore.InterfaceConfig
+}
+
+type GroupMemberStatus struct {
+	group net.IP
+	// localMembers saves the local Pods which have joined in the multicast group. The key is the interface name of the Pod.
+	localMembers map[string]bool
+	lastProbe    time.Time
+	mutex        sync.RWMutex
+}
+
+type Controller struct {
+	ofClient     openflow.Client
+	nodeConfig   *config.NodeConfig
+	igmpSnooper  *IGMPSnooper
+	groupEventCh chan mcastGroupEvent
+	groupCache   cache.Indexer
+	ctx          context.Context
+	cancelFunc   context.CancelFunc
+}
+
+func (m *Controller) Run(stopCh <-chan struct{}) {
+	// Install flows on OVS for IGMP packets and multicast traffic forwarding:
+	// 1) send the IGMP report messages to Antrea Agent,
+	// 2) forward the IMGP query messages to all local Pods,
+	// 3) forward the multicast traffic to antrea-gw0 if no local Pods have joined in the group, and this is to ensure
+	//    local Pods can access the external multicast receivers.
+	m.ofClient.InstallMulticastInitialFlows(uint8(openflow.PacketInReasonMC))
+	// Periodically query Multicast Groups on OVS.
+	go wait.NonSlidingUntil(func() {
+		if err := m.igmpSnooper.queryIGMP(net.IPv4zero, queryVersion); err != nil {
+			klog.ErrorS(err, "Failed to send IGMP query")
+		}
+	}, time.Second*queryIntervalSeconds, stopCh)
+
+	for i := 0; i < workerCount; i++ {
+		// Process multicast Group membership report or leave messages.
+		go m.processGroupEvent(stopCh)
+	}
+
+}
+
+func (m *Controller) processGroupEvent(stopCh <-chan struct{}) {
+	ticker := time.NewTicker(time.Second * queryIntervalSeconds)
+	for {
+		select {
+		case e := <-m.groupEventCh:
+			obj, ok, _ := m.groupCache.GetByKey(e.group.String())
+			switch e.eType {
+			case groupJoin:
+				if !ok {
+					m.addGroupMemberStatus(e)
+				} else {
+					m.updateGroupMemberStatus(obj, e)
+				}
+			case groupLeave:
+				if ok {
+					m.updateGroupMemberStatus(obj, e)
+				}
+			}
+		case <-ticker.C:
+			// Periodically check the group member status, and remove the groups in which no members exist
+			m.clearStaleGroups()
+		case <-stopCh:
+			ticker.Stop()
+			m.cancelFunc()
```

Comment:
The cancelFunc will be called twice. I think you should wait for stopCh in `Run` and then call `m.cancelFunc()` after it.